RFC: transforms

pmeier opened the pull request:
This is an RFC on how transforms will work with the datasets after the rework of the datasets API. Rendered version [here](https://github.com/pmeier/torchvision-datasets-rework/blob/rfc/transforms/transforms.md)

pmeier committed:
initial draft

pmeier commented:
After some more thoughts, I see two additional problems that we need to solve. Happy to hear your ideas / thoughts.

---

What happens if a transform changes some feature attributes for example cropping an image or converting a bounding box to a new format. In the first case the information is at least still encoded in the new shape of the image, but for bounding box conversion, the information is gone. Thus, we need to somehow reflect these changes back to the features.

To achieve that, we could either change the feature dictionary in-place, or we need to somehow return a changed version with each transform.

---

In its current state the `FeatureTransform` has the ability of transforming multiple features, but has no notion of coherence of the features. Thus, it is impossible to share data between individual transformation steps. For example to rotate a bounding box, we need the shape of the underlying image, but there is no way to share it. The same goes for joint random transforms that need to act on the same set of parameters drawn for each sample.

NicolasHug commented on transforms.md:
````diff
@@ -0,0 +1,140 @@
+One pain point for the upstream implementation of transforms is the joint transformation of multiple features with the same transformation. For example images can be rotated with the `Rotate` transform, but corresponding bounding boxes cannot. To overcome this the first proposal of this post is the introduction of a `FeatureTransform` that is able to handle (possibly) multiple `Feature`'s (`Image ...
+
+```python
+class FeatureTransform:
````
Trying to understand the scope: is this class part of a new a public-facing API, or is this an implementation detail that will be hidden to users?

EDIT: reading further down, it looks like it's all user-facing

NicolasHug commented on transforms.md:
````diff
@@ -0,0 +1,140 @@
+```python
+transform = Rotate(30.0)
+transform(Image(), torch.rand(3, 256, 256))
+transform(BoundingBox(type="XYHW"), (0, 0, 256, 256))
````
What is the rational for separating the Feature type and the value?

For example, what is the benefit of this over

```py
transform(Image(torch.rand(3, 256, 256)))
transform(BoundingBox(type="XYHW", (0, 0, 256, 256)))
```

?

NicolasHug commented on transforms.md:
```diff
@@ -0,0 +1,140 @@
+
+One major concern of the upstream implementation of datasets in `torchvision.dataset` are the optional `transform`, `target_transform`, and `transforms`. Apart from not being used in a [standardized way  across datasets](https://gist.github.com/pmeier/14756fe0501287b2974e03ab8d651c10), from a theoretical standpoint the transformation has nothing to do with the dataset.
+
+We already decided that the rework will remove transforms from datasets. That also means we need to provide an alternative. Opposed to the upstream implementation, the new API will return a datapipe and each sample drawn from it will be a dictionary that holds the data. This means we cannot simply `dataset.map(torchvision.transforms.Rotate(30.0))` on since the rotation transformation expects an i ...
```
At least for me, it would greatly help to have a small description of what the proposal below will enable, or what it will make easier to do. Like, if we didn't introduce `FeatureTransform` and `SampleTransform`, what is the code that users would need to write to transform their dataset?

pmeier replied:
I'll add some examples.

NicolasHug reviewed (commented):
Thanks a lot for putting this together @pmeier !

I'm trying to get a better sense of the scope and the challenges here, so I mostly have questions at this point :)

I'll start here with a naive one: instead of introducing the new `SampleTransform` and `FeatureTransform` classes, could we "simply" (<-- gotta love this word) extend our currents transforms from `torchvsion.transforms` to handle not just tensors, but also X where X is what the datapipe returns? This way users could basically do `dataset.map(torchvision.transforms.Rotate(30.0))`, which seems optimal in terms of UX.

(I'm not saying we should do this, but the answer to this question will help me better understand where we're going)

----

Also, some seemingly related discussed sparked in https://github.com/pytorch/vision/issues/4029#issuecomment-870637925. If we start supporting new Tensor-like classes like  `class Image(Tensor)` and `class BoundingBox(Tensor)`, we might be able to avoid the introduction of the new `Feature` structure.

pmeier reviewed (commented):
@NicolasHug 

> Also, some seemingly related discussed sparked in pytorch/vision#4029 (comment). If we start supporting new Tensor-like classes like class Image(Tensor) and class BoundingBox(Tensor), we might be able to avoid the introduction of the new Feature structure.

If we can do that, it would solve a lot of the questions you have and issues I mentioned. PyTorch offers the `__torch_function__` hook. Without going into details, with it it is possible to perform any PyTorch operation, for example transform an image, and return our custom class. So something like

```python
>>> image = torchvision.Image(torch.rand(3, 256, 256))
>>> transformed_image = image + 1
>>> type(transformed_image)
<class 'torchvision.Image'>
```

is possible without any hacks from our side. Doing something like this of course implies that we convert all numerics to tensors.

The problem we had with this approach is that we were unable to come up with a strict enough rule set, that would let us determine if an arbitrary operation *should* return our custom tensor type. Take the example above: is a floating-point tensor with values outside of the range `[0, 1]` still an image? Just to name a second: "If you subtract a random real number from a class label, is the result still a class label?" You can come up with a number of these for all proposed custom tensor classes.

It should be fairly easy to control this for all builtin transformations, but we can't guarantuee that in general.

Thus, we decided to not encode this extra information in the tensor itself. The new datapipes will return a dictionary as sample. Encoding the information in the keys rather than the values is also possible, but this also has two problems:
- Having non-string keys would make working with a sample dictionary a lot more cumbersome.
- We can't guarantuee that a user transformation keeps the keys either, so this has the same problems as encoding the information in the values of the dictionary.

> What is the rational for separating the Feature type and the value?
> 
> For example, what is the benefit of this over
> 
> ```python
> transform(Image(torch.rand(3, 256, 256)))
> transform(BoundingBox(type="XYHW", (0, 0, 256, 256)))
> ```
> 
> ?

Thus, we came up with the plan to encode the extra information in a custom support structure.


IMO in the end this all boils down to two possible scenarios:

1. We provide a more rigid framework that is based on the assumption that the transform ensures that it only returns valid tensor types. This makes the UX better for users that only rely on builtin transforms. At the same time it will be harder to implement custom transformations.
2. We base all of our stuff on very minimal assumptions at the cost of a worse UX.

I vote for option 1., but lets discuss this idea a little more. 

This would also make something like 

> could we "simply" (<-- gotta love this word) extend our currents transforms from `torchvsion.transforms` to handle not just tensors, but also X where X is what the datapipe returns? This way users could basically do `dataset.map(torchvision.transforms.Rotate(30.0))`, which seems optimal in terms of UX.

a lot simpler, because now each transform would know which feature transform to dispatch based on the input type.

pmeier committed:
rewrite proposal to use custom feature classes

pmeier committed:
fix wording

pmeier commented:
After some more thought, I think @NicolasHug's comments hit the nail on the head: my proposal was complicated and still wasn't able to solve that he and I myself pointed out. Thus, I rewrote the proposal with the assumption that we can use custom tensor classes. I've added an implications section to discuss this. Overall I like the new proposal much better. Happy to hear your thoughts.

pmeier committed:
misc test fixes

fmassa commented on torchvision/transforms/_core.py:
```diff
@@ -0,0 +1,83 @@
+        return self._apply_feature_transforms(sample)
+
+    @staticmethod
+    def _apply_image_and_bounding_box_transforms(
```
Mentioned in person, I think if we let `image_size` be a property of `BoundingBox` then this is not needed

pmeier replied:
Yes. If we make each `Feature` self-contained, this can be a lot easier.

fmassa commented on torchvision/features/_core.py:
```diff
@@ -0,0 +1,13 @@
+__all__ = ["Feature"]
+
+
+class Feature(torch.Tensor):
```
This is great!

One meta-question that the team needs to figure out is if we will be supporting torchscript for those features or not. IIUC this might not be propagating the inherited type to torchscripted functions

pmeier replied:
I've split this into `Feature` and `TensorFeature`, since a `Feature` might not necessarily be a `Tensor`. For example, `Text` cannot be represented by a tensor, but could very well be regarded as `Feature` from `torchtext`.

fmassa commented on torchvision/features/_bounding_box.py:
```diff
@@ -0,0 +1,102 @@
+    def format(self) -> BoundingBoxFormat:
+        return self._format
+
+    def convert(self, format: Union[str, BoundingBoxFormat]) -> "BoundingBox":
```
Nice! One other thing to think about is if there would be a minimum common API across all our data types that we might want to enforce?

pmeier replied:
Currently, I don't see anything. Happy to hear ideas though. If something comes up later, that should be retro-fittable.

pmeier replied:
I've added a `TensorFeature.from_tensor` method, that should make it easy to create a new feature from just a tensor.

fmassa reviewed (commented):
This is looking pretty good, thanks a lot @pmeier !

As we've discussed on VC, here are a few notes from our chat.

pmeier committed:
make features self-contained

pmeier committed:
cleanup

pmeier committed:
add compose

pmeier committed:
add auto register and better random support

vmoens commented on transforms.md:
Why not have each transform implementing a default behaviour that can be overloaded for specific features using `__torch_function__`?

pmeier replied:
Not sure if I understand. Is this just about the dispatch? 

Right now the proposal (very simplified) looks like this:

```python
class FooTransform:
    @staticmethod
    def bar(input): ...
    
    @staticmethod
    def baz(input): ...
    
    def forward(self, input):
        if isinstance(input, Bar):
            return self.bar(input)
        elif isinstance(input, Baz):
            return self.baz(input)
        else:
            return input
```

This way we still have a functional API, i.e. `FooTransform.bar` will replace `functional.foo_transform`. Of course it would be even nicer we didn't need to specify the function and could just do `FooTransform.apply` and this would automatically sort out the type. But then, we could also implement that without `__torch_function__`.

Could you provide a snippet how that would look like if we implement what you have in mind?

pmeier replied:
I've added an implementation for my proposed `apply` approach. Basically, we can now do `foo_transform = FooTransform.apply` and also get automatic dispatch in the functional interface:

```python
foo_transform = FooTransform.apply


transformed_image = foo_transform(Image(...))
transformed_bbox = foo_transform(BoundingBox(...))
```

 The only thing left (and I don't know if your approach can solve this) is that the internal registering relies on a naming scheme which is not optimal.

pmeier replied:
This was resolved in an offline discussion. @vmoens proposal was to overload PyTorch ops to behave differently for different input types if needed. One such example is to use `torch.nn.functional.interpolate` with `mode="bilinear"` for `Image`'s and `mode="nearest"` for `SegementationMask`s. 

We indentified three problems with this approach compared to the one presented in this PR:

1. In the example above, only one parameter needs to change, because `Image`'s and `SegmentationMask`s are very closely related and thus the transform almost looks the same. If that is not the case, for example for `Image`'s and `BoundingBoxes`', we would still need some kind of dispatch for the different algorithms.
2. Every time a new transform is contributed, the contributor needs to check if the ops that were used are properly overloaded. This will increase the initial burden, since operator overloading is not an easy concept.
3. Overloading operators hides the actual transform implementation behind some magic, which makes the implementation more opaque.

pmeier committed:
add automatic functional dispatch

pmeier committed:
fix functional interface

pmeier committed:
update proposal

pmeier committed:
introduce from_tensor creation method

vmoens commented on transforms.md:
Why not having a class
```python
class HorizontalFlip(RandomHorizontalFlip):
    def __init__(self):
        super().__init__(p=1.0)
```
That would alleviate the need of coding up twice the same thing but keep the nice UX of having access to a class that is not random.
Would also make scripts clearer.
There is also another way of doing things, which is to have a class `HorizontalFlip` that does the job and a class RandomHorizontalFlip that would just call HorizontalFlip with probability `p` (in other words, it is the opposite version of what i just suggested).
That would also mean that we could have a code that is recycled across transforms.

Something like
```python
class HorizontalFlip(Transform):
    @staticmethod
    def apply_image(image):
        return image.hflip()
```

and 
```python

def random_apply(fun):
    def decorated_fun(self, image, **params):
        p = self.p
        if torch.rand(1).item() > p:
            return fun(self, image, **params)
        return image
    return decorated_fun


class RandomHorizontalFlip(Transform):
    def __init__(self, p):
        super().__init__()
        self.p = p

    @random_apply
    def apply_image(self, image):
        return super().apply_image(image)
```

pmeier replied:
> That would alleviate the need of coding up twice the same thing but keep the nice UX of having access to a class that is not random.

The only issue I see with this is that we need to maintain two sets of docstrings. On the other hand, if we were to apply this pattern to multiple transforms, there is probably some room for automation in this. 

> Would also make scripts clearer.

Yeah, that is a valid concern. Another idea would be to have a `RandomApply` object that wraps a transform, i.e. `transform = RandomApply(Rotate(), p=0.5)`. That should be as clear as having a separate class, but would drastically reduce the maintenance burden.

pmeier committed:
factor out recursive apply

datumbox commented on torchvision/features/_image.py:
```diff
@@ -0,0 +1,11 @@
+__all__ = ["Image"]
+
+
+class Image(TensorFeature):
```
Nice. Maybe on the future that could include information like PIL's `mode` to know the type of the image.

pmeier replied:
I think `mode` is a little overkill, since `PIL` also uses it to differentiate between integral and floating point images. This information is already encoded in the `dtype`. Still, something like `color_space` will be a great addition.

datumbox replied:
I'm not arguing to add it. I said it opens up possibilities.

datumbox commented on torchvision/transforms/_geometry.py:
```diff
@@ -0,0 +1,59 @@
+        return Image(input.flip((-1,)))
+
+    @staticmethod
+    def bounding_box(input: BoundingBox) -> BoundingBox:
```
Agreed. This way, when possible, we share the same functional implementation for images and ~~bboxes~~ masks and when not possible we implement new. Also we leave it up to the user to make the call depending on the use-case rather than assuming the BBox is in some part of the target.

See also this corner case transform that requires joint transform of all image, label, bbox:
https://github.com/pytorch/vision/blob/96f6e0a117d5c56f7e0237851dbb96144ebb110b/references/detection/transforms.py#L54-L129

pmeier replied:
Not sure I understand:

> we share the same functional implementation for images and bboxes and when not possible we implement new

In most cases the implementation will be different for images, bounding boxes and possible other types. In any case, we will always have a separate method for each type. In case the transformation is the same, it is the developers responsibility to implement it with minimal duplications:

```python
class FooTransform(Transform):
    @staticmethod
    def _foo(input):
        return input

    @staticmethod
    def image(input: Image) -> Image:
        return FooTransform._foo(input)

    @staticmethod
    def bounding_box(input: BoundingBox) -> BoundingBox:
        return FooTransform._foo(input)
```

> Also we leave it up to the user to make the call depending on the use-case rather than assuming the BBox is in some part of the target.

That is actually what we are trying to avoid. If you want to, you can call `HorizontalFlip.image(...)`, but during normal operation you would do

```python
transform = HorizontalFlip()
transformed_image = transform(image)
transformed_bbox = transform(bbox)
```

This is needed to be able to transform a complete sample drawn from a dataset, which might include more than just an image. 

For a better overview of what we are trying to achieve, please read the [accompanying document](https://github.com/pmeier/torchvision-datasets-rework/blob/rfc/transforms/transforms.md) which is updated as this RFC progresses.

datumbox replied:
@pmeier I didn't disagree with you. I just said I like the approach :)

What I meant is that there are a few transforms that can be implemented for images and bboxes in the same way. That can live in the functional API and as you said, we can call it for both `image()` and `bounding_box()`. But also the proposed API supports cases when this is not the case by providing separate implementations for the two.

> That is actually what we are trying to avoid. 

Yes I had a discussion with @fmassa who explained that you want to avoid this to make things composable. Sounds good. Just make sure you look the example I sent you because there are transforms where image, bbox and labels must be processed together. I think the API can cover for that but it's worth having this use-case in mind.

datumbox commented on torchvision/transforms/_geometry.py:
```diff
@@ -0,0 +1,59 @@
+        self._dist = torch.distributions.Uniform(low, high)
+
+    def get_params(self, sample: Any = None) -> Dict[str, Any]:
+        return dict(degrees=self._dist.sample().item())
```
Returning all the random parameters in a single function call will likely not work for iterative algorithms such as:
https://github.com/pytorch/vision/blob/96f6e0a117d5c56f7e0237851dbb96144ebb110b/references/detection/transforms.py#L81-L129

The same might apply for data augmentation techniques such as [UniformAugment](https://arxiv.org/abs/2003.14348) where there is a random number of looping through transforms that themselves have random parameters. Example:
https://github.com/automl/trivialaugment/blob/master/aug_lib.py#L672-L677

In such cases, it might be hard to have a Functional form that does not have side-effects due to calling random.

vmoens replied:
very interesting and challenging point. So if I understand well, the only way to get around those use cases is to have a transform design that takes all objects to be transformed through the same function at the same time, as they might need to be able to 'see' each other. Am I right?

pmeier replied:
That seems to be the case. This should be doable in the current approach, since `get_params` receives the full sample and thus has access to all the information.

fmassa commented on torchvision/transforms/_core.py:
```diff
@@ -0,0 +1,109 @@
+                continue
+            argspec = inspect.getfullargspec(value)
+
+            # transform takes none or more than one than one positional arguments
+            if len(argspec.args) != 1:
+                continue
```
As we discussed in VC, I believe this can lead to subtle bugs. If the user writes their transform as
```python
class Rotate(Transform):
    @staticmethod
    def image(img, angle):
        pass
```
instead of
```python
class Rotate(Transform):
    @staticmethod
    def image(img, *, angle):
        pass
```
then the transformation will not be automatically registered, and no errors / warnings will show up, only that the transform will silently not be applied because it wasn't properly auto-registered.

I think it might be better to be explicit here and instead use decorators to register a function to a transform. Something in the lines of
```python
class Rotate(Transform):
    pass

@register_transform(transform=Rotate, input_type=Image)
def rotate_image(img, angle):
    pass
```

pmeier replied:
> I think it might be better to be explicit here and instead use decorators to register a function to a transform.

1. I dislike this approach, because we would have to manually namespace everything with the feature type. For example, I think it is more concise to have `Rotate.image` and `Rotate.bounding_box` rather than `rotate_image` and `rotate_bounding_box`. Note that in the latter case, the namespace would have roughly n times more entries where n is the number of features we support.
2. The auto registering can be disabled and the registering can be performed manually in the constructor:

    ```python
    class Rotate(Transform, auto_register=False):
        def __init__(self):
            super().__init__()
            self.register_feature_transform(Image, self.foo)
    
        @staticmethod
        def foo(input, angle):
            pass
    ```
3. If we ultimately go with registering standalone functions with a decorator, I suggest we use a classmethod of the transform, i.e.

    ```python
    class Rotate(Transform):
        pass
    
    
    @Rotate.register(Image)
    def rotate_image(input, angle):
        pass
    ```

> As we discussed in VC, I believe this can lead to subtle bugs.

True, good catch! The rule can be made more robust though. We evoke the feature transforms with `feature_transform(input, **params)` where `params` can be an empty dict. That means all of these signatures need to be supported:

```python
def foo(input):
    pass

def foo(input=None):
    pass

def foo(input, bar):  # your example
    pass

def foo(input, bar=None):
    pass

def foo(input=None, bar=None):
    pass

def foo(input, *, bar):
    pass

def foo(input, *, bar=None):
    pass

def foo(input=None, *, bar):
    pass

def foo(input=None, *, bar=None):
    pass
```

For Python `>=3.8` we also need to keep positional-only arguments in mind. 

For the debugging, the user can use `transform.is_supported`(EDIT: as of writing this comment, this method only existed locally. It is included in the most recent commit) and pass an feature type or instance to see if it would be transformed or not. But I agree that one would probably only do that after discovering that something is wrong. Another idea would be to add a `verbose` flag that prints extra information during the auto registration.

---

Since all of this is a trade-off between convenience with possible subtle bugs and explicitness with a higher verbosity, I think we should wait for more opinions before we decide.

pmeier replied:
I've added a commit that makes the matching more robust and adds the option to print information about the process:

```python
class TestTransform(Transform, verbose=True):
    @staticmethod
    def incompatible_signature1():
        pass

    @staticmethod
    def incompatible_signature2(*, input):
        pass

    # Python >=3.8 only
    # @staticmethod
    # def incompatible_signature3(input, foo, /):
    #     pass

    @staticmethod
    def _private(input):
        pass

    @staticmethod
    def unknown(input):
        pass

    @staticmethod
    def imaeg(input):
        pass

    @staticmethod
    def image(input, foo):
        pass

    @staticmethod
    def boundingbox(input):
        pass

    @staticmethod
    def bounding_box(input, *, foo):
        pass
```

```
TestTransform._private() was not registered as feature transform, because it is private.
TestTransform.bounding_box() was registered as feature transform for type 'BoundingBox'.
TestTransform.boundingbox() was not registered as feature transform, because its name doesn't match any known feature type. Did you mean to name it 'bounding_box' to be registered for type 'BoundingBox'?
TestTransform.imaeg() was not registered as feature transform, because its name doesn't match any known feature type. Did you mean to name it 'image' to be registered for type 'Image'?
TestTransform.image() was registered as feature transform for type 'Image'.
TestTransform.incompatible_signature1() was not registered as feature transform, because it cannot be invoked with incompatible_signature1(input, **params).
TestTransform.incompatible_signature2() was not registered as feature transform, because it cannot be invoked with incompatible_signature2(input, **params).
TestTransform.unknown() was not registered as feature transform, because its name doesn't match any known feature type.
```

fmassa commented on torchvision/features/_bounding_box.py:
```diff
@@ -0,0 +1,160 @@
+            self.formats.XYWH: self._xyxy_to_xywh,
+            self.formats.CXCYWH: self._xyxy_to_cxcywh,
+        }
+
```
For when we will be discussing specifics of the data classes, it might be good to add a `__torch_function__` that forbids applying operations of two bounding boxes if they have different `image_size`, while allowing `BoundingBox` + `Tensor` to work

pmeier committed:
improve auto registering

pmeier committed:
add READMe on how to implement a new transform

pmeier committed:
add implementation for MixUp

ppwwyyxx commented:
It seems a big issue of the current prototype is that it cannot implement `Compose.get_params`, therefore cannot implement nested Compose.
* Why it cannot implement `Compose.get_params` (efficiently):
   `Compose.get_params` cannot be implemented simply by calling each sub-transform's `get_params`.  For example let's try to obtain `Compose.get_params` of `Compose([A, B, C])`. To call `B.get_params` you need the input sample of B. Therefore you need to first __run__ the transform of A to transform the sample, before you can call `B.get_params`. 
   In other words, __getting the params__ already requires __running/applying the transform__. When the params are returned they are used again (in `Transform.forward()`) to apply the transform - which is a 2x waste of computation.

It may seem nested compose is useless because it can be flattened, but that's not the case. Beyond `Compose` as a sequential container, there might be other types of containers, such as `Compose(RandomOrder([T1, T2, T3]), T4)`, `RandomChoice([Compose([T1, T2]), T3])`. These pipelines that have nested structures will all have this issue.

pmeier commented:
@ppwwyyxx 

> It seems a big issue of the current prototype is that it cannot implement `Compose.get_params`, therefore cannot implement nested Compose.

I agree with the argument, but disagree with the conclusion. 

During "normal" operation the transform containers (`Compose`, `RandomChoice`, ...) do not need access to the `get_params` method. Every transformation simply receives the sample it is supposed to transform and thus there is no duplicated computation. Example: Assume we have `Compose[Compose[T1, T2], T3]`. `T1` receives the `sample`, `T2` receives `T1(sample)`, and `T3` receives `Compose[T1, T2](sample)`. 

Still, there might be scenarios where you conclusion is right.

If we have two equal but separate transforms that should operate with the same parameters:

```python
transform1 = MyTransform()
transform2 = MyTransform()
params = transform1.get_params()
transformed_image1 = transform1(image, params=params)
transformed_image2 = transform(image, params=params)
```

I currently fail to see a use case for this, but maybe there is one.

pmeier committed:
add support for creating transforms from callables and other transforms

pmeier committed:
add better support for containers

pmeier committed:
add documentation

pmeier commented:
@ppwwyyxx I've added minimal implementations for `RandomChoice` and `RandomApply`.

pmeier commented:
With the latest commits, the [accompanying document](https://github.com/pmeier/torchvision-datasets-rework/blob/rfc/transforms/transforms.md) is no longer in sync. I'll update it ASAP.

pmeier committed:
update RFC document

pmeier committed:
improve query

pmeier committed:
add example how feature transform requirements can look like

pmeier committed:
add minimal implementation for segmentation training

pmeier committed:
refactorget_params to be able to return other parameters based on type

pmeier committed:
simplify transform wrapping

pmeier committed:
add RandomOrder container

ppwwyyxx commented:
Thanks @pmeier for following up! I agree that you can implement containers by giving up the ability to do `get_params`. However I do feel that having `get_params` available is pretty important.  I'll list some scenarios where this is useful:


1. With `get_params` users can save the params and later use them to apply an inverse of the transforms (if the transform supports it). This is useful during inference as well as test-time augmentation, where we transform the images, detect objects on the new image, and then need to invert the transform so we map the detected boxes/masks back to the original image space.

2. Users can obtain a sequence of params, and apply the transforms one by one. This allows them to insert logic in between. With code this looks like:
```
# transforms: a composite transform container (e.g. with Composite, RandomOrder, etc)
tfms_and_prms: List[Tuple[Transform, dict]] = transforms.get_params(inputs)
for tfm, param in tfms_and_prms:
   inputs = tfm(inputs, params=param)
   # HERE: do some extra work with inputs
```
Insertion of this logic have a few use cases, e.g.:
* For bounding boxes, it's up to users whether they want to clip the boxes to image sizes after every transform (might be needed after rotation/affine). They can make this decision at here
* For keypoints, users need to modify the keypoints labels (e.g. left eye -> right eye) if transform is a flip. Users can implement this logic if they can obtain parameters of the transform, to check if a flip has happened.
* Keypoints in COCO has a "visibility" field, users may (or may not) want to check after every transform whether some keypoints are out of range, and mark them as invisible.

These extra but very simple logic can be added in the above snippet without changing the input types transforms will see. 
Without the ability to manually apply the params, the above custom logic seem still doable, but will probably need to be added as new data types or subclasses of existing data types, and then register new transform methods for the types. I'm not sure at the moment how easy it is to extend the types but maybe that's not bad.

However, actually the above snippet is fundamentally difficult to realize, because as I commented earlier doing `get_params` already needs to run the transforms so we waste some computation. The way we address this in detectron2 is to separately define "a minimal subset of inputs that are needed to get_params" which we called "AugInput". This also feels a bit awkward.. so I don't have a good suggestion on what exactly to do. But I hope the above example use cases can be useful.

pmeier commented:
@ppwwyyxx Thanks a lot for your input!

> 1. With `get_params` users can save the params and later use them to apply an inverse of the transforms (if the transform supports it). This is useful during inference as well as test-time augmentation, where we transform the images, detect objects on the new image, and then need to invert the transform so we map the detected boxes/masks back to the original image space.

I agree, that is an important use case that should be supported. Would it be sufficient to have a functional inverse

```python
transform = MyComplexTransform()

sample = next(dataset)

transformed_sample = transform(sample)
transformed_prediction = model(sample["input"])

# magically materialize this here, more about that later
params = ...
prediction = transform.inverse(transformed_prediction, params=params)
```

or should we actually instantiate an inverse transform? 

> inverse of the transforms (if the transform supports it)

How do you decide if a transform has an inverse or not? For example, `crop` is the inverse of `pad`, but is `pad` also considered an inverse to `crop`? I would say no for images, since `pad` cannot retrieve the information lost by `crop`. For bounding boxes or keypoints on the other hand I would say yes, since no information was lost here.

> 2. Users can obtain a sequence of params, and apply the transforms one by one. This allows them to insert logic in between.

Although I generally dislike polymorphism, it can maybe make things simpler here. We could have a `return_params: bool = False` flag in the forward such that 

```python
transform = MyComplexTransform()

sample = next(dataset)

transformed_sample1, params = transform(sample, return_params=True)
transformed_sample2 = transform(sample, params=params)

assert transformed_sample1 == transformed_sample2
```

This would transform your snippet into

```python
# transforms: a composite transform container (e.g. with Composite, RandomOrder, etc)
for tfm in transforms:
    inputs, params = tfm(inputs)
    # HERE: do some extra work with inputs
```

and avoid extracting the params upfront either with doubled computation or dummy inputs.

What do you think?
